File: queue/Python/Sliding Window Maximum.py
# Problem link: https://leetcode.com/problems/sliding-window-maximum/
# Time: O(n)
# Space: O(n)
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        queue, left, result = deque(), 0, []

        for right in range(len(nums)):
            while queue and queue[-1] < nums[right]:
                queue.pop()
            
            queue.append(nums[right])

            if right - left + 1 == k:
                result.append(queue[0])

                if queue[0] == nums[left]:
                    queue.popleft()
                
                left += 1
                
        return result

File: queue/Java/Sliding Window Maximum.java
// Problem link: https://leetcode.com/problems/sliding-window-maximum/
// Time: O(n)
// Space: O(n)
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int[] result = new int[nums.length - k + 1];
        Deque<Integer> queue = new LinkedList<>();
        int left = 0, idx = 0;

        for(int right = 0; right < nums.length; right++) {

            while(!queue.isEmpty() && queue.peekLast() < nums[right])
                queue.pollLast();
            
            queue.offerLast(nums[right]);

            if(right - left + 1 == k) {
                result[idx++] = queue.peekFirst();

                if(nums[left] == queue.peekFirst()) {
                    queue.pollFirst();
                }
                left++;
            }
        }

        return result;

    }
}
